src/Cornelis/Highlighting.hs

```diff
@@ -74,82 +76,73 @@ highlightBuffer b hs = do
   pure $ concat holes
 
 newtype LineIntervals = LineIntervals
-  { li_intervalMap :: IntervalMap BufferOffset (LineNumber, Text)
+  { li_intervalMap :: IntervalMap AgdaIndex (LineNumber 'ZeroIndexed, Text)
     -- ^ Mapping from positions to line numbers
   }
   deriving newtype (Semigroup, Monoid)
 
-
 getLineIntervals :: Vector Text -> LineIntervals
-getLineIntervals = LineIntervals . go (Offset 0) (LineNumber 0)
+getLineIntervals = LineIntervals . go (oneIndexed 1) (zeroIndexed 0)
   where
     go
-        :: BufferOffset
-        -> LineNumber
+        :: AgdaIndex
+        -> LineNumber 'ZeroIndexed
         -> Vector Text
-        -> IntervalMap BufferOffset (LineNumber, Text)
-    go (Offset pos) line v
+        -> IntervalMap AgdaIndex (LineNumber 'ZeroIndexed, Text)
+    go pos line v
       | Just (t, ss) <- V.uncons v =
         let len = T.length t
-            pos' = pos + fromIntegral len
-        in IM.singleton (IM.Interval (coerce pos) $ coerce pos') (line, t)
-              <> go (coerce $ pos' + 1) (incLineNumber line) ss
+            pos' = pos .+ Offset len
+        in IM.insert (IM.Interval pos pos') (line, t)
+              $ go (incIndex pos') (incIndex line) ss
       | otherwise = mempty
 
-lookupPoint :: LineIntervals -> BufferOffset -> Maybe (LineNumber, Int64)
-lookupPoint (LineIntervals im) off = do
-  (IM.Interval scs _, (startline, s)) <- listToMaybe $ IM.search off im
-  pure ( startline
-       , fromIntegral $ toBytes s (offsetDiff off scs) - 1
-       )
-
+lookupPoint :: LineIntervals -> AgdaIndex -> Maybe VimPos
+lookupPoint (LineIntervals im) i = do
+  (IM.Interval lineStart _, (line, s)) <- listToMaybe $ IM.search i im
+  let col = toBytes s (zeroIndexed 0 .+ (i .-. lineStart))
+  pure (Pos line col)
 
 ------------------------------------------------------------------------------
 -- | Returns any holes it tried to highlight on the left
 addHighlight
     :: Buffer
     -> LineIntervals
     -> Highlight
-    -> Neovim CornelisEnv ([Interval' Int64], Maybe Extmark)
+    -> Neovim CornelisEnv ([VimInterval], Maybe Extmark)
 addHighlight b lis hl = do
-  case (,)
-          <$> lookupPoint lis (hl_start hl)
-               -- Subtract 1 here from the end offset because unlike everywhere
```

Comment:
How do you deal with extmarks' ranges? This comment and the next suggest extmarks are treated differently than intervals everywhere else in vim.

Reply:
That's fair, though the docs do clearly say it's inclusive: https://neovim.io/doc/user/api.html#api-indexing

> Exception: the following API functions use [extmarks](https://neovim.io/doc/user/api.html#extmarks) indexing (0-based indices, end-inclusive):

I don't know though! So long as it looks good and the tests pass I'm happy.

Reply:
Ah that's confusing. My reference was what it says for the specific function `nvim_buf_set_extmark` https://neovim.io/doc/user/api.html#nvim_buf_set_extmark()

> end_row : ending line of the mark, 0-based inclusive.  
> end_col : ending col of the mark, 0-based exclusive.

Reply:
What an awful tool we've decided to implement support for :)